### ISSUE TITLE
Confusing error message: "deref: invalid pointer"

### ISSUE THREAD

**Issue by facundominguez** [well-typed/visualize-cbn#1]
```
$ cat t.hs
span = (\p -> \xs ->
  case xs of {
   Nil -> Pair xs xs ;
   Cons x xs' ->
     if p x then
      let res = @span p xs'
      in Pair (case @res of { Pair ys zs -> Cons x ys})
              (case @res of { Pair ys zs -> zs })
     else Pair Nil xs
  }
)

main = case (@span (\x -> True) (Cons 1 (Cons 2 (Cons 3 Nil)))) of {
  Pair ys zs -> case ys of {
    Cons y ys2 -> ys ;
    Nil -> ys
    }
  }
```

```
$ git log -1
commit d8a1ebc0a29d68b5dde8d8fbadd6c85bc47c6a7b
Author: Edsko de Vries <edsko@well-typed.com>
Date:   Fri Sep 15 17:27:40 2017 +0200

    Add README
$ .cabal-sandbox/bin/visualize-cbn -i t.hs --show-trace --hide-prelude
...
visualize-cbn: deref: invalid pointer Ptr Nothing (Just "res")
CallStack (from HasCallStack):
  error, called at src/CBN/Heap.hs:78:26 in main:CBN.Heap
```

**Comment by edsko** [comment]
So the error message certainly leaves something to be desired here :) But you should write

```
      let res = @span p xs'
      in Pair (case res of ..
```

The `@` notation refers to heap pointers. In a `let` statement the variable isn't a pointer yet (not until the `let` is executed and space is allocated on the heap). In the initial code only references to CAFs (i.e., top-level definitions) are allocated on the heap.

But I should definitely clarify this and this error message. Thanks for the report!

**Comment by facundominguez** [comment]
Alright, that fixes it indeed. Thanks.

**Comment by edsko** [comment]
This got fixed in #9 , the error message is now

```
(stuck: Invalid reference to symbol @res. Valid symbol names are: ["@span","p","xs","ys"])
```
